### ISSUE TITLE
Quotas are not being calculated and checked properly

### ISSUE THREAD

**Issue by JustinKyleJames** [irods/irods#4089]
### iRODS Version, OS and Version

iRODS 4.2.4
Tested on Ubuntu 14.04

### Observed behavior (including steps to reproduce, if applicable)

I see two separate issues:

1. When a quota is set at multiple levels (resource and total) the quota check passes if any quota is not exceeded.  It should fail if any quota is exceeded.

2.  If you set a quota on a coordinating resource, the "iadmin cu" does not update the usage on that coordinating resource.

**Comment by JustinKyleJames** [comment]
There are a couple of issues I noticed in the code.  

1.  In plugins/database/src/general_query.cpp, in generateSpecialQuery() the queries return a list ordered by quota_over descending.  This ordering is correct.  However, in queRescQuota() in server/api/src/rsGetRescQuota.cpp, a link list is being created from bottom to top so the first one on the list is the one with the lowest quota_over.  Then in setRescQuota() it just takes the first one from the list.

2.  The actual queries return information for all resources.  That makes me believe even if the issue above was fixed, it would still not work because we shouldn't care about the quota on a resource that is not being written to.  These queries need to be rewritten.

**Comment by trel** [comment]
Is this still true?   Do we not have tests around this behavior?

**Comment by korydraughn** [comment]
We have https://github.com/irods/irods/blob/ad59ae9c52331ddbc19dac71bc05461b00f46885/scripts/irods/test/test_quotas.py.

**Comment by korydraughn** [comment]
I don't think there are any other tests around quotas.

**Comment by trel** [comment]
Okay, so confirmed, there is nothing in that file that tests when both types of quotas are in play at the same time.

**Comment by korydraughn** [comment]
> 2. If you set a quota on a coordinating resource, the "iadmin cu" does not update the usage on that coordinating resource.

Just one comment to add following discussion about this.

Q. Should physical quotas be allowed on coordinating resources?
A. We feel it should. Whether that's a good use of the feature is up to the admin.

---

The question is motivated by a resource hierarchy containing a replication resource and three unixfilesystem resources. If a quota is attached to the replication resource and a user stores a file there, the replication resource will replicate it to its children.

What should the quota report for the replication resource?
Should the user's file cost 3x its size due to there being three physical replicas?

This situation led us to the answer of ... the quota system is about physical storage and should count all replicas. The quota system is a tool and admins are given freedom over where it is applied.

**Comment by trel** [comment]
Agreed - I have created a separate issue for the coordinating resource total/rollup - #8667.

This issue should now only be about confirming/testing the existing math for group quotas when there are multiple quotas in play.

**Comment by alanking** [comment]
Checked boxes. @FifthPotato - Please close if complete. Thanks

**Comment by FifthPotato** [comment]
> Checked boxes. [@FifthPotato](https://github.com/FifthPotato) - Please close if complete. Thanks

Requesting an un-check for now. The next test case I'm going for might also be covered under this issue:

> If you set a quota on a coordinating resource, the "iadmin cu" does not update the usage on that coordinating resource.

**Comment by trel** [comment]
coordinating is now at https://github.com/irods/irods/issues/8667

i think this can be closed.

**Comment by alanking** [comment]
I am seeing a failure in `test_quotas.Test_Quotas.test_physical_resource_and_total_quota_simultaneously__issue_4089` on Ubuntu 24.04 / Postgres 16:
```python
File "/var/lib/irods/scripts/irods/test/test_quotas.py", line 194, in test_physical_resource_and_total_quota_simultaneously__issue_4089
    self.admin.assert_icommand(['iadmin', 'lq'], 'STDOUT', ['\nquota_over: -5000\n', '\nquota_over: -10904\n'])
File "/var/lib/irods/scripts/irods/test/session.py", line 162, in assert_icommand
    return assert_command(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/var/lib/irods/scripts/irods/test/command.py", line 76, in assert_command
    return _assert_helper(*args, should_fail=False, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/var/lib/irods/scripts/irods/test/command.py", line 104, in _assert_helper
    assert result
           ^^^^^^
AssertionError
```

Here is the failing assertion (note the "quota_over: **-10903**" line, where **-10904** was expected):
```
 --- IrodsSession: icommand executed by [otherrods#tempZone] [iadmin lq] --- 
Assert Command: iadmin lq
Expecting STDOUT: ['\nquota_over: -5000\n', '\nquota_over: -10904\n']
  stdout:
    | Per resource quotas (if any) for users/groups:
    | user_name: test_group_case_40_41_42_1
    | zone_name: tempZone
    | resc_name: ufs0_quota_case_40_41_42
    | quota_limit: 5000
    | quota_over: -5000
    | modify_ts: 2025-11-07.17:40:49
    | 
    | Global (total usage) quotas (if any) for users/groups:
    | user_name: test_group_case_40_41_42_1
    | zone_name: tempZone
    | quota_limit: 15000
    | quota_over: -10903
    | modify_ts: 2025-11-07.17:40:50
  stderr:
    | 
Output not found

FAILED TESTING ASSERTION
```
Here's what I ran in the testing environment:
```
python3 run_core_tests.py --project-dir projects/ubuntu-24.04/ubuntu-24.04-postgres-16/ --irods-package-dir ~/hdd/builds/irods_packages/main/ubuntu-24.04 -v --concurrent 4
```

**Comment by alanking** [comment]
@FifthPotato - Please close if complete. Thanks
